libc-bottom-half/sources/chdir.c

```diff
@@ -0,0 +1,144 @@
+#include <errno.h>
+#include <fcntl.h>
+#include <limits.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+#include <wasi/libc-find-relpath.h>
+#include <wasi/libc.h>
+
+#ifdef _REENTRANT
+#error "chdir doesn't yet support multiple threads"
+#endif
+
+extern char *__wasilibc_cwd;
+static int __wasilibc_cwd_mallocd = 0;
+
+int chdir(const char *path)
+{
+    static char *relative_buf = NULL;
+    static size_t relative_buf_len = 0;
+
+    // Find a preopen'd directory as well as a relative path we're anchored
+    // from which we're changing directories to.
+    const char *abs;
+    int parent_fd = __wasilibc_find_relpath_alloc(path, &abs, &relative_buf, &relative_buf_len, 1);
+    if (parent_fd == -1)
+        return -1;
+
+    // Make sure that this directory we're accessing is indeed a directory.
+    struct stat dirinfo;
+    int ret = fstatat(parent_fd, relative_buf, &dirinfo, 0);
+    if (ret == -1)
+        return -1;
+    if (!S_ISDIR(dirinfo.st_mode)) {
+        errno = ENOTDIR;
+        return -1;
+    }
+
+    // Copy over our new absolute path into `__wasilibc_cwd`. Only copy over
+    // the relative portion of the path if it's not equal to `.`
+    size_t len = strlen(abs) + 1;
+    int copy_relative = strcmp(relative_buf, ".") != 0;
+    char *new_cwd = malloc(len + (copy_relative ? strlen(relative_buf) : 0));
+    if (new_cwd == NULL) {
+        errno = ENOMEM;
+        return -1;
+    }
+    new_cwd[0] = '/';
+    strcpy(new_cwd + 1, abs);
+    if (copy_relative)
+        strcpy(new_cwd + 1 + strlen(abs), relative_buf);
```

Comment:
It looks like this needs a `/` between the `abs` part and the `relative_buf` part. Does `relative_buf` always start with a `/` here? If so, it'd be good to comment that somewhere. If not, this should add a `/` between the paths.

Reply:
Ah yes indeed, and I've fixed up another issue or two with this.